masonry/src/passes/layout.rs

```diff
@@ -0,0 +1,196 @@
+    // If it's true, that's probably an error.
+    state.needs_layout = false;
+
+    state.local_paint_rect = state
+        .local_paint_rect
+        .union(new_size.to_rect() + state.paint_insets);
+
+    #[cfg(debug_assertions)]
+    {
+        for child_id in widget.children_ids() {
+            let child_id = child_id.to_raw();
+            let child_state_mut = state_mut
+                .children
+                .get_child_mut(child_id)
+                .unwrap_or_else(|| panic!("widget #{child_id} not found"));
+            let child_state = child_state_mut.item;
+            if child_state.is_expecting_place_child_call {
+                debug_panic!(
+                        "Error in '{}' #{}: missing call to place_child method for child widget '{}' #{}. During layout pass, if a widget calls WidgetPod::layout() on its child, it then needs to call LayoutCtx::place_child() on the same child.",
+                        widget.short_type_name(),
+                        id,
+                        child_state.widget_name,
+                        child_state.id.to_raw(),
+                    );
+            }
+
+            // TODO - This check might be redundant with the code updating local_paint_rect
+            let child_rect = child_state.paint_rect();
+            if !rect_contains(&state.local_paint_rect, &child_rect) && !state.is_portal {
+                debug_panic!(
+                        "Error in '{}' #{}: paint_rect {:?} doesn't contain paint_rect {:?} of child widget '{}' #{}",
+                        widget.short_type_name(),
+                        id,
+                        state.local_paint_rect,
+                        child_rect,
+                        child_state.widget_name,
+                        child_state.id.to_raw(),
+                    );
+            }
+        }
+    }
+
+    // TODO - Figure out how to deal with the overflow problem, eg:
+    // What happens if a widget returns a size larger than the allowed constraints?
+    // Some possibilities are:
+    // - Always clip: might be expensive
+    // - Display it anyway: might lead to graphical bugs
+    // - Panic: too harsh?
+    // Also, we need to avoid spurious crashes when we initialize the app and the
+    // size is (0,0)
+    // See https://github.com/linebender/xilem/issues/377
+
+    let state_mut = parent_ctx
+        .widget_state_children
+        .get_child_mut(id)
+        .expect("WidgetPod: inner widget not found in widget tree");
+    parent_ctx.widget_state.merge_up(state_mut.item);
+    state_mut.item.size = new_size;
+
+    log_layout_issues(widget.short_type_name(), new_size);
+
+    true
+}
+
+fn log_layout_issues(type_name: &str, size: Size) {
+    if size.width.is_infinite() {
+        warn!("Widget `{type_name}` has an infinite width.");
+    }
+    if size.height.is_infinite() {
+        warn!("Widget `{type_name}` has an infinite height.");
+    }
+}
+
+pub(crate) fn run_layout_on<W: Widget>(
```

Comment:
We should have a small doc comment here. E.g.:
```suggestion
/// Run [`Widget::layout`] method on the widget contained in `pod`.
/// This will be called by [`LayoutCtx::run_layout`], which is itself called in the parent widget's `layout`.
pub(crate) fn run_layout_on<W: Widget>(
```